experimental/oak_functions_loader_linux_vsock/src/channel.rs

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::proto::oak::sandbox::runtime::{Request, Response};
+use anyhow::anyhow;
+use byteorder::{BigEndian, ByteOrder};
+use prost::Message;
+use ringbuf::{Consumer, Producer, RingBuffer};
+use std::{
+    io::{Cursor, Read, Write},
+    mem::size_of,
+};
+
+const BUFFER_SIZE: usize = 65536;
+
+// Implements a Bedebox communication channel encoding in which every message is
+// represented as a `length:value` where `value` is a Protobuf encoded message.
+pub struct Channel<T>
+where
+    T: Read + Write
+{
+    stream: T,
+    read_buffer_producer: Producer<u8>,
+    read_buffer_consumer: Consumer<u8>,
+}
+
+impl<T> Channel<T>
+where
+    T: Read + Write
+{
+    pub fn new(stream: T) -> Self {
+        let read_buffer = RingBuffer::<u8>::new(BUFFER_SIZE);
+        let (read_buffer_producer, read_buffer_consumer) = read_buffer.split();
+
+        Self {
+            stream,
+            read_buffer_producer,
+            read_buffer_consumer,
+        }
+    }
+}
+
+impl<T> oak_baremetal_communication_channel::Read for Channel<T>
+where
+    T: std::io::Read + std::io::Write
+{
+    fn read(&mut self, data: &mut [u8]) -> anyhow::Result<()> {
+        let mut buffer = vec![0; BUFFER_SIZE];
+        let read_bytes = self
+            .stream
+            .read(&mut buffer)
+            .map_err(|error| anyhow!("Couldn't read from stream: {:?}", error))?;
+        if read_bytes > 0 {
+            let mut current_buffer = &buffer[..read_bytes];
+            let mut cursor = Cursor::new(&mut current_buffer);
+
+            // Read multiple size:value pairs from the buffer.
+            while !cursor.is_empty() {
```

Comment:
Why is the buffer/cursor needed? This copies data from the stream to a buffer and then reads from the buffer using a cursor so that is looks like a stream again. Why not just read direclty from the stream in the first place?